addon_imps/storage/s3.py

```diff
@@ -0,0 +1,112 @@
+        s3 = boto3.client(
+            "s3", aws_access_key_id=access_key, aws_secret_access_key=secret_key
+        )
+        try:
+            s3.list_buckets()
+        except exceptions.ClientError:
+            raise ValidationError("Fail to validate access key and secret key")
+
+    @sync_to_async
+    def construct_client(self, account):
+        access_key = account._credentials.decrypted_credentials.access_key
+        secret_key = account._credentials.decrypted_credentials.secret_key
+        self.client = boto3.client(
+            "s3", aws_access_key_id=access_key, aws_secret_access_key=secret_key
+        )
+
+    async def get_item_info(self, item_id: str) -> storage.ItemResult:
+        bucket = item_id.split("/", 1)[0]
+        key = item_id.split("/", 1)[1]
+        response = self.client.list_objects(Bucket=bucket, Prefix=key, Delimiter="/")
+        if response["Contents"]:
+            if len(response["Contents"]) == 1 and ("CommonPrefixes" not in response):
+                # that means this is a file, not a folder
+                return storage.ItemResult(
+                    item_id=response["Contents"][0]["Key"],
+                    item_name=response["Contents"][0]["Key"],
+                    item_type=storage.ItemType.FILE,
+                )
+            else:
+                # That means this is a folder
+                return storage.ItemResult(
+                    item_id=item_id,
+                    item_name=item_id,
+                    item_type=storage.ItemType.FOLDER,
+                )
+        return None
+
+    async def list_root_items(self, page_cursor: str = "") -> storage.ItemSampleResult:
+        results = list(self.list_buckets())
+        return storage.ItemSampleResult(
+            items=results,
+            total_count=len(results),
+        )
+
+    async def list_child_items(
+        self,
+        item_id: str,
+        page_cursor: str = "",
+        item_type: storage.ItemType | None = None,
+    ) -> storage.ItemSampleResult:
+        bucket = item_id.split("/", 1)[0]
+        key = item_id.split("/", 1)[1]
+        response = self.client.list_objects(Bucket=bucket, Prefix=key, Delimiter="/")
+        results = []
+        if response["CommonPrefixes"]:
+            for folder in response["CommonsPrefixes"]:
+                results.append(
+                    storage.ItemResult(
+                        item_id=folder["Prefix"],
+                        item_name=folder["Prefix"],
+                        item_type=storage.ItemType.FOLDER,
+                    )
+                )
+        if response["Contents"]:
+            for file in response["Contents"]:
+                results.append(
+                    storage.ItemResult(
+                        item_id=file["Key"],
+                        item_name=bucket["Key"],
+                        item_type=storage.ItemType.FILE,
+                    )
+                )
+        return
```

Comment:
If we can distinguish between an id that doesn't exist and an id that has no child items, then we should raise error when the id doesn't exist, and we should return an empty list when the ide has no child items.